File: java/medium/TheyAreEverywhere.java
// https://codeforces.com/problemset/problem/701/C

package java.medium;

import java.io.*;
import java.util.*;

public class TheyAreEverywhere {

    public static void main(String[] args)throws Exception {

        BufferedReader br =new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        String s = br.readLine();

        HashSet<Character> set = new HashSet<>();

        for(char c:s.toCharArray())
            set.add(c);

        int total=set.size();

        HashMap<Character,Integer> map = new HashMap<>();

        int left=0;
        int ans=n;

        for(int right=0;right<n;right++){

            char ch=s.charAt(right);

            map.put(ch,map.getOrDefault(ch,0)+1);

            while(map.size()==total){

                ans=Math.min(ans,right-left+1);

                char c = s.charAt(left);

                map.put(c,map.get(c)-1);

                if(map.get(c)==0)
                    map.remove(c);

                left++;
            }
        }

        System.out.println(ans);
    }
}

File: c++/medium/they_are_everywhere.cpp
#include<bits/stdc++.h>
using namespace std;

int main(){

    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin>>n;

    string s;
    cin>>s;

    set<char> st(
    s.begin(),
    s.end());

    int total=st.size();

    unordered_map<char,int> freq;

    int left=0;
    int ans=n;

    for(int right=0;
        right<n;
        right++){

        freq[s[right]]++;

        while(
        freq.size()==total){

            ans=min(
            ans,
            right-left+1);

            freq[
            s[left]
            ]--;

            if(
            freq[
            s[left]
            ]==0)

            freq.erase(
            s[left]);

            left++;
        }
    }

    cout<<ans;
}